byway/src/main/webapp/routepage.js

```diff
@@ -56,42 +70,75 @@ function initMap() {
  * @param {String} end ending point of route. TODO: In the final implementation, just have 
  * start as start and end are the same
  */
-function calcRoute(directionsService, directionsRenderer, start, end) {
+
+function calcRoute() {
   let request = {
     origin:  start,
     destination: end,
-    travelMode: 'DRIVING'
+    travelMode: 'DRIVING',
+    waypoints: Array.from(waypoints),
+    optimizeWaypoints: true
   };
   directionsService.route(request, function(response, status) {
     if (status == 'OK') {
       directionsRenderer.setDirections(response);
+      computeTotalDistance(response);
     } else {
       window.alert("Could not calculate route due to: " + status);
     }
   });
 }
 
+/**
+ * Calculates and sums up the distance and time duration between all destinations (legs)
+ * @param {response} result response from the directions service object
+ */
+function computeTotalDistance(result) {
+  let totalDist = 0;
+  let totalTime = 0;
+  // full route
+  let route = result.routes[0];
+  for (let i = 0; i < route.legs.length; i++) {
+    // in meters
+    totalDist += route.legs[i].distance.value;
+    // in seconds
+    totalTime += route.legs[i].duration.value;
+  }
+
+  totalDist = (totalDist / 1000).toFixed(2);
+  let hours = Math.floor(totalTime / 3600);
+  let minutes = Math.round((totalTime - hours*3600) / 60);
+ 
+  document.getElementById("distance").innerText = totalDist + "km";
+  document.getElementById("duration").innerText = 
+    (hours == 0) ? minutes + " mins" : hours + " hr " + minutes + " mins";
+}
```

Comment:
Instead of writing these directly to DOM, consider separating calculation from display by having this method return a plain object with the computed values, and then writing that to dom as a separate operation.

Reply:
Okay, I've separated them into two functions. Please let me know if you had something else in mind.

Reply:
When doing a pure computation, it's best to avoid "side effects", that is modifying something other than the data you're working on  (or modifying the data you're given at all!). In this case, you have a function which could compute the length of the trip without causing any other effects, except for that you call the function to update the DOM at the bottom. A function which has no side-effects is more generally useful than one with side-effects, because a function without side-effects can be used from anywhere without worrying about what it might do to the page. So like, if you decide you need the length of the trip for some other computation, you could call the function without worrying that it also updates the page. This is why I say to return the computed value as a plain object, like `return {totalDist, hours, minutes};` rather than directly calling into code that updates the DOM. Then whever called the functio ncan decide what to do with that data, wheter that's update the DOM or use it as part of a different calculation.

Reply:
Ohh I see what you mean. I changed my implementation to this.